### PR TITLE
Support fallbacks configuration with wildcards, e.g. 'fr-*': 'fr'

### PR DESCRIPTION
**Pull request by kclang**

### CODE REVIEW

**Review by tuanlc — APPROVED** [review]

File: i18n.js [hunk 1]
```diff
@@ -804,9 +783,60 @@ module.exports = (function() {
...
-   * searches for locale in given object
+   * Setup fallbacks for a given language.
+   * @param {Array} acceptedLanguages  languages from the HTTP Accept-Language header
+   * @param {JSON} fallbacks           fallbacks option
+   * @param {String} lang              current language
+   * @param {String} parentLang        current parent language
+   * @param {JSON} wildCardChecked     flags check if fallback for current language has been set up
    */
+  var setupLanguageFallbacks = function(acceptedLanguages, fallbacks, lang, parentLang,
+    wildCardChecked) {
```
Comment (sngo25) [inline i18n.js]: bad indent

File: i18n.js [hunk 2]
```diff
@@ -804,9 +783,60 @@ module.exports = (function() {
   };
 
   /**
-   * searches for locale in given object
+   * Setup fallbacks for a given language.
+   * @param {Array} acceptedLanguages  languages from the HTTP Accept-Language header
+   * @param {JSON} fallbacks           fallbacks option
```
Comment (sngo25) [inline i18n.js]: should be Object instead of JSON

File: i18n.js [hunk 3]
```diff
@@ -721,36 +720,16 @@ module.exports = (function() {
       // 'accept-language' is the most common source
       if (languageHeader) {
         var acceptedLanguages = getAcceptedLanguagesFromHeader(languageHeader),
-          match, fallbackMatch, fallback;
+          match, fallbackMatch;
+
+        var wildCardChecked = {};
```
Comment (sngo25) [inline i18n.js]: what do you use this var for? It seems to be useless

File: i18n.js [hunk 4]
```diff
@@ -804,9 +783,60 @@ module.exports = (function() {
...
+        wildCardChecked[parentLang] = true;
+      }
+      // Check if we have a configured fallback set for the parent language of the locale.
+      if (fallbacks[parentLang]) {
+        fallback = fallbacks[parentLang];
+        // Fallbacks for a parent language should be inserted
+        // to the end of the list, so they're only picked
+        // if there is no better match.
+        if (acceptedLanguages.indexOf(fallback) < 0) {
+          acceptedLanguages.push(fallback);
```
Comment (sngo25) [inline i18n.js]: this function should not modify the parameters (AKA making side effect), it should return something

File: i18n.js [hunk 5]
```diff
@@ -804,9 +783,60 @@ module.exports = (function() {
   };
 
   /**
-   * searches for locale in given object
+   * Setup fallbacks for a given language.
```
Comment (sngo25) [inline i18n.js]: this description is not useful enough

**Review by sngo25 — CHANGES_REQUESTED** [review]

File: i18n.js (hunk 3 above)
Comment (kclang) [reply i18n.js]: It's used in setupLanguageFallbacks function

File: i18n.js (hunk 1 above)
Comment (kclang) [reply i18n.js]: it's good as usual

File: i18n.js (hunk 4 above)
Comment (kclang) [reply i18n.js]: re-structured code, please re-check

File: i18n.js (hunk 2 above)
Comment (kclang) [reply i18n.js]: done

File: i18n.js [hunk 6]
```diff
@@ -720,57 +717,12 @@ module.exports = (function() {
...
-          }
-          if (region) {
-            regions.push(region.toLowerCase());
-          }
-
-          if (!match && locales[lang]) {
-            match = lang;
-            break;
-          }
+        var bestMatch = getBestMatchLanguage(acceptedLanguages);
```
Comment (sngo25) [inline i18n.js]: bestMatchLanguage

**Review by sngo25 — CHANGES_REQUESTED** [review]

File: i18n.js (hunk 6 above)
Comment (kclang) [reply i18n.js]: changed

**Review by sngo25 — APPROVED** [review]

**Pull request merged**